server/leafnode.go

```diff
@@ -2527,11 +2527,15 @@ func (s *Server) leafNodeFinishConnectProcess(c *client) {
 	c.mu.Unlock()
 
 	// Make sure we register with the account here.
-	c.registerWithAccount(acc)
-	s.addLeafNodeConnection(c, _EMPTY_, _EMPTY_, false)
-	s.initLeafNodeSmapAndSendSubs(c)
-	if sendSysConnectEvent {
-		s.sendLeafNodeConnect(acc)
+	if err := c.registerWithAccount(acc); err != nil {
+		c.Errorf("Registering leaf with account %s resulted in error: %v", acc.Name, err)
+		c.closeConnection(ProtocolViolation)
```

Comment:
Have a return after this call, and remvoe the `else` condition below. The cleanup being done at the end of this function was in case issues occur between the successful registration and adding to the server map, etc.. here since you fail early, there is no need to do what is at the end, so simply return.

Reply:
Thanks for pointing that out. fixed